File: app/src/main/java/eu/darken/sdmse/common/forensics/csi/dalvik/DalvikProfileCSI.java
//package eu.darken.sdmse.common.forensics.csi.dalvik;
//
//import java.io.File;
//import java.util.ArrayList;
//import java.util.Collection;
//import java.util.List;
//
//import androidx.annotation.NonNull;
//import androidx.annotation.Nullable;
//import eu.thedarken.sdm.App;
//import eu.thedarken.sdm.tools.ApiHelper;
//import eu.thedarken.sdm.tools.forensics.FileForensics;
//import eu.thedarken.sdm.tools.forensics.Location;
//import eu.thedarken.sdm.tools.forensics.LocationInfo;
//import eu.thedarken.sdm.tools.forensics.OwnerInfo;
//import eu.thedarken.sdm.tools.forensics.csi.CSICheck;
//import eu.thedarken.sdm.tools.forensics.csi.CSIProcessor;
//import eu.thedarken.sdm.tools.forensics.csi.checks.ClutterCheck;
//import eu.thedarken.sdm.tools.io.SDMFile;
//import eu.thedarken.sdm.tools.storage.Storage;
//
//public class CSIDalvikProfile extends CSIProcessor {
//    static final String TAG = App.logTag("CSIDalvikProfile");
//    private final List<CSICheck> checks = new ArrayList<>();
//
//    public CSIDalvikProfile(FileForensics fileForensics) {
//        super(fileForensics);
//        checks.add(new DirNameCheck(this));
//        checks.add(new ClutterCheck(this));
//    }
//
//    @Override
//    public boolean hasJurisdiction(@NonNull Location type) {
//        return type == Location.DALVIK_PROFILE;
//    }
//
//    @Nullable
//    @Override
//    public LocationInfo matchLocation(@NonNull SDMFile file) {
//        if (ApiHelper.hasLolliPop()) {
//            Collection<Storage> storages = getStorageManager().getStorages(Location.DALVIK_PROFILE, true);
//            for (Storage storage : storages) {
//                String base = storage.getFile().getPath() + File.separator;
//                if (file.getPath().startsWith(base)) {
//                    return new LocationInfo(file, Location.DALVIK_PROFILE, base, true, storage);
//                }
//            }
//        }
//        return null;
//    }
//
//    @Override
//    public void process(@NonNull OwnerInfo ownerInfo) {
//        for (CSICheck check : checks) {
//            if (check.check(ownerInfo)) return;
//        }
//    }
//}

File: app/src/main/java/eu/darken/sdmse/common/forensics/csi/dalvik/DalvikProfileCSI.kt
package eu.darken.sdmse.common.forensics.csi.dalvik

import dagger.Binds
import dagger.Module
import dagger.Reusable
import dagger.hilt.InstallIn
import dagger.hilt.components.SingletonComponent
import dagger.multibindings.IntoSet
import eu.darken.sdmse.common.areas.DataArea
import eu.darken.sdmse.common.areas.DataAreaManager
import eu.darken.sdmse.common.areas.currentAreas
import eu.darken.sdmse.common.debug.logging.logTag
import eu.darken.sdmse.common.files.APath
import eu.darken.sdmse.common.files.isAncestorOf
import eu.darken.sdmse.common.forensics.AreaInfo
import eu.darken.sdmse.common.forensics.CSIProcessor
import eu.darken.sdmse.common.forensics.csi.LocalCSIProcessor
import eu.darken.sdmse.common.forensics.csi.dalvik.tools.DalvikClutterCheck
import eu.darken.sdmse.common.forensics.csi.dalvik.tools.DirNameCheck
import javax.inject.Inject

@Reusable
class DalvikProfileCSI @Inject constructor(
    private val areaManager: DataAreaManager,
    private val dirNameCheck: DirNameCheck,
    private val clutterCheck: DalvikClutterCheck,
) : LocalCSIProcessor {

    override suspend fun hasJurisdiction(type: DataArea.Type): Boolean = type == DataArea.Type.DALVIK_PROFILE

    override suspend fun identifyArea(target: APath): AreaInfo? = areaManager.currentAreas()
        .filter { it.type == DataArea.Type.DALVIK_PROFILE }
        .mapNotNull { area ->
            if (!area.path.isAncestorOf(target)) return@mapNotNull null

            AreaInfo(
                dataArea = area,
                file = target,
                prefix = area.path,
                isBlackListLocation = true
            )
        }
        .singleOrNull()

    override suspend fun findOwners(areaInfo: AreaInfo): CSIProcessor.Result {
        val results = mutableSetOf<DalvikCheck.Result>()
        results.add(dirNameCheck.process(areaInfo))

        if (results.all { it.owners.isEmpty() }) {
            results.add(clutterCheck.process(areaInfo))
        }

        return CSIProcessor.Result(
            owners = results.map { it.owners }.flatten().toSet(),
            hasKnownUnknownOwner = results.any { it.hasKnownUnknownOwner },
        )
    }

    @Module @InstallIn(SingletonComponent::class)
    abstract class DIM {
        @Binds @IntoSet abstract fun mod(mod: DalvikProfileCSI): CSIProcessor
    }

    companion object {
        val TAG: String = logTag("CSI", "Dalvik", "Profile")
    }
}